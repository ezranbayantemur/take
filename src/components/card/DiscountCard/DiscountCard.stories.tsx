import React from 'react';
import type {Meta, StoryObj} from '@storybook/react-native';
import {DiscountCard} from '@components';
import {DiscountCardProps} from './DiscountCard.types';

const DiscountCardMeta: Meta<DiscountCardProps> = {
  title: 'DiscountCard',
  component: DiscountCard,
};

export default DiscountCardMeta;

export const Basic: StoryObj<DiscountCardProps> = {
  decorators: [
    () => {
      return (
        <>
          <DiscountCard
            testID=""
            data={{
              categoryTitle: 'electronics',
              totalPrice: 1000,
              discountPercentage: 5,
              discountedPrice: 950,
              hasDiscount: true,
              hasProduct: true,
              remainingPricetoApplyDiscount: 0,
            }}
          />
          <DiscountCard
            testID=""
            data={{
              categoryTitle: 'jewelry',
              totalPrice: 1000,
              discountPercentage: 17,
              discountedPrice: 0,
              hasDiscount: false,
              hasProduct: true,
              remainingPricetoApplyDiscount: 1500,
            }}
          />
        </>
      );
    },
  ],
};
