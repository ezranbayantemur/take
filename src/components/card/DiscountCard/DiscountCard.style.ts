import {color, font, padding} from '@style';
import {StyleSheet} from 'react-native';

export default StyleSheet.create({
  container: {
    margin: padding.MEDIUM,
  },
  discount_container: {
    paddingVertical: padding.MEDIUM,
    borderRadius: 5,
    padding: padding.SMALL,
    backgroundColor: color.MONZA,
  },
  not_enough_discount_container: {
    paddingVertical: padding.MEDIUM,
    borderRadius: 5,
    padding: padding.SMALL,
    backgroundColor: color.MERCURY,
  },
  discount_text: {
    color: color.ALABASTER,
    fontFamily: font.REGULAR,
  },
  discount_new_price_text: {
    fontFamily: font.BLACK,
  },
  not_enough_discount_text: {
    fontFamily: font.REGULAR,
  },
});
