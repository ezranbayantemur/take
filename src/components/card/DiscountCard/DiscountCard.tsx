import {View, Text} from 'react-native';
import React from 'react';
import styles from './DiscountCard.style';
import type {DiscountCardProps} from './DiscountCard.types';

const categoryTitleParser: Record<string, string> = {
  jewelry: 'Takı',
  electronic: 'Elektronik',
  men_clothing: 'Erkek Giyim',
  women_clothing: 'Kadın Giyim',
};

const DiscountCard = ({
  testID,
  data: {
    categoryTitle,
    discountPercentage,
    discountedPrice,
    hasDiscount,
    hasProduct,
    remainingPricetoApplyDiscount,
    totalPrice,
  },
}: DiscountCardProps) => {
  if (!hasProduct) {
    return null;
  }

  const parsedCategoryTitle = categoryTitleParser[categoryTitle];
  if (hasDiscount) {
    return (
      <View testID={`${testID}_discount_discountcard`} style={styles.container}>
        <View style={styles.discount_container}>
          <Text style={styles.discount_text}>
            {parsedCategoryTitle} kategorisindeki ürünlere %{discountPercentage}{' '}
            indirim uygulandı. Yeni toplu fiyat {totalPrice} TL yerine{' '}
            <Text style={styles.discount_new_price_text}>
              {discountedPrice} TL
            </Text>{' '}
          </Text>
        </View>
      </View>
    );
  }

  return (
    <View
      testID={`${testID}_not_enough_discount_discountcard`}
      style={styles.container}>
      <View style={styles.not_enough_discount_container}>
        <Text style={styles.not_enough_discount_text}>
          {parsedCategoryTitle} kategorisinde %{discountPercentage} indirim
          fırsatını yakalamak için bu kategoriden{' '}
          <Text style={styles.not_enough_discount_remain_price_text}>
            {remainingPricetoApplyDiscount} TL
          </Text>
          'lik daha ürün sepete ekleyin
        </Text>
      </View>
    </View>
  );
};

export default DiscountCard;
