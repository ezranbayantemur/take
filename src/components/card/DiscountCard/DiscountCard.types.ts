import {Discount} from '@types';

export interface DiscountCardProps {
  testID: string;
  data: Discount;
}
