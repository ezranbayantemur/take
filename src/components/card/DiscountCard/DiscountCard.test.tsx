import React from 'react';
import {render} from '@testing-library/react-native';
import DiscountCard from './DiscountCard';

let wrapper: ReturnType<typeof render>;

describe('DiscountCard unit tests', () => {
  beforeEach(() => {
    wrapper = render(
      <DiscountCard
        testID="test"
        data={{
          categoryTitle: 'electronics',
          totalPrice: 1000,
          discountPercentage: 5,
          discountedPrice: 950,
          hasDiscount: true,
          hasProduct: true,
          remainingPricetoApplyDiscount: 0,
        }}
      />,
    );
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render null if there is no product for discount category', () => {
    wrapper.rerender(
      <DiscountCard
        testID="test"
        data={{
          categoryTitle: 'electronics',
          totalPrice: 0,
          discountPercentage: 5,
          discountedPrice: 0,
          hasDiscount: true,
          hasProduct: false,
          remainingPricetoApplyDiscount: 0,
        }}
      />,
    );

    const component = wrapper.queryByTestId('test_discountcart');
    expect(component).toBeNull();
  });

  it('should render correct discount text if there is a discount to apply', () => {
    wrapper.rerender(
      <DiscountCard
        testID="test"
        data={{
          categoryTitle: 'electronics',
          totalPrice: 1000,
          discountPercentage: 5,
          discountedPrice: 950,
          hasDiscount: true,
          hasProduct: true,
          remainingPricetoApplyDiscount: 0,
        }}
      />,
    );
    const text = wrapper.queryByText(
      'Elektronik kategorisindeki ürünlere %5 indirim uygulandı. Yeni toplu fiyat 1000 TL yerine 950 TL',
    );
    expect(text).not.toBeNull();
  });

  it('should render correct discount text if there not enough total sub to apply discount', () => {
    wrapper.rerender(
      <DiscountCard
        testID="test"
        data={{
          categoryTitle: 'jewelry',
          totalPrice: 1000,
          discountPercentage: 10,
          discountedPrice: 0,
          hasDiscount: false,
          hasProduct: true,
          remainingPricetoApplyDiscount: 1500,
        }}
      />,
    );

    const text = wrapper.queryByText(
      "Takı kategorisinde %10 indirim fırsatını yakalamak için bu kategoriden 1500 TL'lik daha ürün sepete ekleyin",
    );
    expect(text).not.toBeNull();
  });
});
