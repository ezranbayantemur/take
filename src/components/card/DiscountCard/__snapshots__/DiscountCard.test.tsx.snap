// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiscountCard unit tests should render correctly 1`] = `
<View
  style={
    {
      "margin": 16,
      "marginVertical": 8,
    }
  }
  testID="test_discount_discountcard"
>
  <View
    style={
      {
        "backgroundColor": "#DB040E",
        "borderRadius": 5,
        "padding": 8,
        "paddingVertical": 16,
      }
    }
  >
    <Text
      style={
        {
          "color": "#FBFBFB",
          "fontFamily": "Lato-Regular",
        }
      }
    >
       kategorisindeki ürünlere %
      5
       
      indirim uygulandı. Yeni toplu fiyat 
      1000
       TL yerine
       
      <Text
        style={
          {
            "fontFamily": "Lato-Black",
          }
        }
      >
        950
         TL
      </Text>
       
    </Text>
  </View>
</View>
`;
